File: vllm_doctor/clients/prometheus.py
import math

import httpx

from vllm_doctor.clients._http import _get
from vllm_doctor.clients.exceptions import ClientConnectionError, ClientError, ClientQueryError
from vllm_doctor.clients.models import MetricSample, label_selector

__all__ = ["PrometheusClient", "ClientConnectionError", "ClientError", "ClientQueryError"]


class PrometheusClient:
    def __init__(
        self,
        base_url: str,
        timeout: float = 10.0,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self._client = client or httpx.AsyncClient(timeout=timeout)

    async def query(self, metric_name: str, time: str | None = None) -> list[MetricSample]:
        params: dict[str, str] = {"query": metric_name}
        if time is not None:
            params["time"] = time

        response = await _get(self._client, f"{self.base_url}/api/v1/query", params=params)
        data = response.json()
        if data.get("status") != "success":
            raise ClientQueryError(data.get("error", "unknown error"))

        return [
            MetricSample(
                labels=r["metric"],
                value=float(r["value"][1]),
                timestamp=float(r["value"][0]),
            )
            for r in data["data"]["result"]
        ]

    async def query_range(self, metric_name: str, start: str, end: str, step: str) -> list[MetricSample]:
        params = {"query": metric_name, "start": start, "end": end, "step": step}

        response = await _get(self._client, f"{self.base_url}/api/v1/query_range", params=params)
        data = response.json()
        if data.get("status") != "success":
            raise ClientQueryError(data.get("error", "unknown error"))

        return [
            MetricSample(
                labels=r["metric"],
                value=float(point[1]),
                timestamp=float(point[0]),
            )
            for r in data["data"]["result"]
            for point in r["values"]
        ]

    async def query_increase(self, metric_name: str, since: str) -> list[MetricSample]:
        return await self.query(f"increase({metric_name}[{since}])")

    async def query_percentile(
        self, metric: str, quantile: float, model: str | None = None, since: str = "5m"
    ) -> float | None:
        sel = label_selector(model)
        expr = f"histogram_quantile({quantile}, sum by (le) (rate({metric}_bucket{sel}[{since}])))"
        samples = await self.query(expr)
        if not samples or not math.isfinite(samples[0].value):
            return None
        return samples[0].value

    async def aclose(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> "PrometheusClient":
        return self

    async def __aexit__(self, *args: object) -> None:
        await self.aclose()

File: rust/vllm-doctor/src/clients/prometheus.rs
//! Prometheus query API client.
use reqwest::Client as HttpClient;

use super::error::ClientError;
use super::{Client, label_selector};
use crate::metrics::series::MetricSample;

#[derive(Debug)]
pub struct PrometheusClient {
    base_url: String,
    client: HttpClient,
}

impl PrometheusClient {
    pub fn new(base_url: impl Into<String>, timeout: f64) -> Result<Self, ClientError> {
        Self::with_client(
            base_url,
            HttpClient::builder()
                .timeout(std::time::Duration::from_secs_f64(timeout))
                .build()?,
        )
    }

    pub fn with_client(
        base_url: impl Into<String>,
        client: HttpClient,
    ) -> Result<Self, ClientError> {
        Ok(Self {
            base_url: base_url.into().trim_end_matches('/').to_string(),
            client,
        })
    }

    async fn get(
        &self,
        url: &str,
        params: &[(String, String)],
    ) -> Result<serde_json::Value, ClientError> {
        let response = self.client.get(url).query(params).send().await?;
        response.error_for_status_ref()?;
        Ok(response.json().await?)
    }

    fn parse_samples(result: &[serde_json::Value]) -> Vec<MetricSample> {
        result
            .iter()
            .filter_map(|r| {
                let labels = r
                    .get("metric")?
                    .as_object()?
                    .iter()
                    .map(|(k, v)| (k.clone(), v.as_str().unwrap_or_default().to_string()))
                    .collect();
                let value = r.get("value")?.get(1)?.as_str()?.parse().ok()?;
                let timestamp = r.get("value")?.get(0)?.as_f64();
                Some(MetricSample {
                    labels,
                    value,
                    timestamp,
                })
            })
            .collect()
    }
}

#[async_trait::async_trait]
impl Client for PrometheusClient {
    async fn query(&self, metric_name: &str) -> Result<Vec<MetricSample>, ClientError> {
        let url = format!("{}/api/v1/query", self.base_url);
        let params = [("query".to_string(), metric_name.to_string())];
        let data = self.get(&url, &params).await?;

        if data.get("status").and_then(|v| v.as_str()) != Some("success") {
            return Err(ClientError::Query(
                data.get("error")
                    .and_then(|v| v.as_str())
                    .unwrap_or("unknown error")
                    .to_string(),
            ));
        }

        let result = data
            .get("data")
            .and_then(|d| d.get("result"))
            .and_then(|r| r.as_array())
            .cloned()
            .unwrap_or_default();
        Ok(Self::parse_samples(&result))
    }

    async fn query_increase(
        &self,
        metric_name: &str,
        since: &str,
    ) -> Result<Option<Vec<MetricSample>>, ClientError> {
        let expr = format!("increase({metric_name}[{since}])");
        let samples = self.query(&expr).await?;
        Ok(Some(samples))
    }

    async fn query_percentile(
        &self,
        metric: &str,
        quantile: f64,
        model: Option<&str>,
        since: &str,
    ) -> Result<Option<f64>, ClientError> {
        let sel = label_selector(model);
        let expr = format!(
            "histogram_quantile({quantile}, sum by (le) (rate({metric}_bucket{sel}[{since}])))"
        );
        let samples = self.query(&expr).await?;
        Ok(samples.into_iter().next().and_then(|s| {
            if s.value.is_finite() {
                Some(s.value)
            } else {
                None
            }
        }))
    }
}
